### PR TITLE
WIP: Periodic dg problems

### PR DESCRIPTION
**Pull request by eric-c-cyr**
@trilinos/panzer

Included for inspection...not ready to merge!

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [ ] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by eric-c-cyr** [comment]
@rppawlo and @krcb

**Comment by krcb** [comment]
@eric-c-cyr @rppawlo It doesn't look like I'm authorized to checkout this PR (I'm not listed as a collaborator on this project).

**Comment by krcb** [comment]
I was able to access this as a patch file, which I downloaded and tried to build. I see link errors of the form:

```
Linking CXX executable packages/panzer/adapters-stk/example/PoissonInterfaceExample/PanzerAdaptersSTK_PoissonInterfaceExample.exe
FAILED: packages/panzer/adapters-stk/example/PoissonInterfaceExample/PanzerAdaptersSTK_PoissonInterfaceExample.exe 

<excessively long link line removed>

Undefined symbols for architecture x86_64:
  "panzer::reorder(std::__1::vector<int, std::__1::allocator<int> >&, std::__1::function<void (int, int)>)", referenced from:
      panzer::IntegrationValues2<double>::generateSurfaceCubatureValues(PHX::MDField<double, panzer::Cell, panzer::BASIS, panzer::Dim, void, void, void, void, void> const&) in libpanzer-disc-fe.a(Panzer_IntegrationValues2.cpp.o)
      panzer::IntegrationValues2<Sacado::Fad::DFad<double> >::generateSurfaceCubatureValues(PHX::MDField<Sacado::Fad::DFad<double>, panzer::Cell, panzer::BASIS, panzer::Dim, void, void, void, void, void> const&) in libpanzer-disc-fe.a(Panzer_IntegrationValues2.cpp.o)
ld: symbol(s) not found for architecture x86_64
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
```

where `<excessively long link line>` is

```
: && /Users/kbeckwi/Documents/Codes/spack/opt/spack/darwin-sierra-x86_64/clang-4.0.1/mpich-3.2.1-bglmbhxgt4wildlhu7moo5cgtp2xbqbn/bin/mpicxx  -g -O2 -std=c++11 -fopenmp=libomp -O3 -DNDEBUG -Wl,-search_paths_first -Wl,-headerpad_max_install_names  packages/panzer/adapters-stk/example/PoissonInterfaceExample/CMakeFiles/PanzerAdaptersSTK_PoissonInterfaceExample.dir/main.cpp.o  -o packages/panzer/adapters-stk/example/PoissonInterfaceExample/PanzerAdaptersSTK_PoissonInterfaceExample.exe  packages/panzer/adapters-stk/src/libpanzer-stk.a packages/muelu/adapters/libmuelu-adapters.a packages/muelu/src/Interface/libmuelu-interface.a packages/muelu/src/libmuelu.a packages/teko/src/libteko.a packages/ifpack2/adapters/libifpack2-adapters.a packages/ifpack2/src/libifpack2.a packages/seacas/libraries/ioss/src/main/libio_info_lib.a packages/seacas/libraries/ioss/src/init/libIonit.a packages/seacas/libraries/ioss/src/transform/libIotr.a packages/seacas/libraries/ioss/src/heartbeat/libIohb.a packages/seacas/libraries/ioss/src/generated/libIogn.a packages/seacas/libraries/ioss/src/visualization/libIovs.a packages/seacas/libraries/ioss/src/pamgen/libIopg.a packages/seacas/libraries/ioss/src/exo_fac/libIoexo_fac.a packages/seacas/libraries/ioss/src/exo_fpp/libIofx.a packages/seacas/libraries/ioss/src/exodus/libIoex.a packages/seacas/libraries/ioss/src/libIoss.a packages/seacas/libraries/exodus/libexodus.a liblast_lib.a packages/panzer/disc-fe/src/libpanzer-disc-fe.a packages/panzer/dof-mgr/src/libpanzer-dof-mgr.a packages/phalanx/src/libphalanx.a packages/panzer/core/src/libpanzer-core.a packages/piro/src/libpiro.a packages/muelu/adapters/libmuelu-adapters.a packages/muelu/src/Interface/libmuelu-interface.a packages/muelu/src/libmuelu.a packages/intrepid2/src/libintrepid2.a packages/sacado/src/libsacado.a packages/tempus/src/libtempus.a packages/rythmos/src/librythmos.a packages/nox/src-loca/src-thyra/liblocathyra.a packages/nox/src-loca/src-epetra/liblocaepetra.a packages/nox/src-loca/src-lapack/liblocalapack.a packages/nox/src-loca/src/libloca.a packages/nox/src-epetra/libnoxepetra.a packages/nox/src-lapack/libnoxlapack.a packages/nox/src/libnox.a packages/teko/src/libteko.a packages/ifpack2/adapters/libifpack2-adapters.a packages/ifpack2/src/libifpack2.a packages/zoltan2/src/libzoltan2.a packages/xpetra/sup/libxpetra-sup.a packages/xpetra/src/libxpetra.a packages/anasazi/tpetra/src/libanasazitpetr:src/pamgen/libIopg.a packages/seacas/libraries/ioss/src/exo_fac/libIoexo_fac.a packages/seacas/libraries/ioss/src/exo_fpp/libIofx.a packages/seacas/libraries/ioss/src/exodus/libIoex.a packages/seacas/libraries/ioss/src/libIoss.a packages/seacas/libraries/exodus/libexodus.a liblast_lib.a packages/panzer/disc-fe/src/libpanzer-disc-fe.a packages/panzer/dof-mgr/src/libpanzer-dof-mgr.a packages/phalanx/src/libphalanx.a packages/panzer/core/src/libpanzer-core.a packages/piro/src/libpiro.a packages/muelu/adapters/libmuelu-adapters.a packages/muelu/src/Interface/libmuelu-interface.a packages/muelu/src/libmuelu.a packages/intrepid2/src/libintrepid2.a packages/sacado/src/libsacado.a packages/tempus/src/libtempus.a packages/rythmos/src/librythmos.a packages/nox/src-loca/src-thyra/liblocathyra.a packages/nox/src-loca/src-epetra/liblocaepetra.a packages/nox/src-loca/src-lapack/liblocalapack.a packages/nox/src-loca/src/libloca.a packages/nox/src-epetra/libnoxepetra.a packages/nox/src-lapack/libnoxlapack.a packages/nox/src/libnox.a packages/teko/src/libteko.a packages/ifpack2/adapters/libifpack2-adapters.a packages/ifpack2/src/libifpack2.a packages/zoltan2/src/libzoltan2.a packages/xpetra/sup/libxpetra-sup.a packages/xpetra/src/libxpetra.a packages/anasazi/tpetra/src/libanasazitpetra.a packages/anasazi/epetra/util/ModeLaplace/libModeLaplace.a packages/anasazi/epetra/src/libanasaziepetra.a packages/anasazi/src/libanasazi.a packages/stk/stk_io/stk_io/util/libstk_io_util.a packages/stk/stk_io/stk_io/libstk_io.a packages/seacas/libraries/ioss/src/main/libio_info_lib.a packages/seacas/libraries/ioss/src/init/libIonit.a packages/seacas/libraries/ioss/src/transform/libIotr.a packages/seacas/libraries/ioss/src/heartbeat/libIohb.a packages/seacas/libraries/ioss/src/generated/libIogn.a packages/seacas/libraries/ioss/src/visualization/libIovs.a packages/seacas/libraries/ioss/src/pamgen/libIopg.a packages/seacas/libraries/ioss/src/exo_fac/libIoexo_fac.a packages/seacas/libraries/ioss/src/exo_fpp/libIofx.a packages/seacas/libraries/ioss/src/exodus/libIoex.a packages/seacas/libraries/ioss/src/libIoss.a packages/seacas/libraries/exodus/libexodus.a /Users/kbeckwi/Documents/Codes/spack/opt/spack/darwin-sierra-x86_64/clang-4.0.1/netcdf-4.4.1.1-66exy2ubcaw3j7oybcr76uxiuggrd45w/lib/libnetcdf.dylib packages/pamgen/src/libpamgen_extras.a packages/pamgen/src/libpamgen.a packages/stk/stk_mesh/stk_mesh/base/libstk_mesh_base.a packages/shards/src/libshards.a packages/stk/stk_topology/stk_topology/libstk_topology.a packages/stk/stk_util/stk_util/use_cases/libstk_util_use_cases.a packages/stk/stk_util/stk_util/registry/libstk_util_registry.a packages/stk/stk_util/stk_util/diag/libstk_util_diag.a packages/stk/stk_util/stk_util/environment/libstk_util_env.a packages/stk/stk_util/stk_util/parallel/libstk_util_parallel.a packages/stk/stk_util/stk_util/util/libstk_util_util.a /Users/kbeckwi/Documents/Codes/spack/opt/spack/darwin-sierra-x86_64/clang-4.0.1/boost-1.66.0-djrx2idkjzpjfyii22fc4cpxnjiyzf3o/lib/libboost_program_options.dylib /Users/kbeckwi/Documents/Codes/spack/opt/spack/darwin-sierra-x86_64/clang-4.0.1/boost-1.66.0-djrx2idkjzpjfyii22fc4cpxnjiyzf3o/lib/libboost_system.dylib packages/stratimikos/src/libstratimikos.a packages/stratimikos/adapters/belos/src/libstratimikosbelos.a packages/stratimikos/adapters/aztecoo/src/libstratimikosaztecoo.a packages/stratimikos/adapters/amesos/src/libstratimikosamesos.a packages/stratimikos/adapters/ml/src/libstratimikosml.a packages/stratimikos/adapters/ifpack/src/libstratimikosifpack.a packages/ml/src/libml.a packages/ifpack/src/libifpack.a packages/amesos/src/libamesos.a packages/common/auxiliarySoftware/SuiteSparse/src/libtrilinosss.a packages/belos/tpetra/src/libbelostpetra.a packages/belos/epetra/src/libbelosepetra.a packages/belos/src/libbelos.a packages/thyra/adapters/tpetra/src/libthyratpetra.a packages/thyra/adapters/epetraext/src/libthyraepetraext.a packages/epetraext/src/libepetraext.a /Users/kbeckwi/Documents/Codes/spack/opt/spack/darwin-sierra-x86_64/clang-4.0.1/hdf5-1.10.1-fyvd5il7akdg4ijgz5fiz3jm5ubid5wd/lib/libhdf5.dylib /usr/lib/libz.dylib /Users/kbeckwi/Documents/Codes/spack/opt/spack/darwin-sierra-x86_64/clang-4.0.1/hdf5-1.10.1-fyvd5il7akdg4ijgz5fiz3jm5ubid5wd/lib/libhdf5_hl.dylib packages/thyra/adapters/epetra/src/libthyraepetra.a packages/thyra/core/src/libthyracore.a packages/rtop/src/librtop.a packages/tpetra/core/ext/libtpetraext.a packages/tpetra/core/inout/libtpetrainout.a packages/tpetra/core/src/libtpetra.a packages/kokkos-kernels/src/libkokkoskernels.a packages/kokkos/algorithms/src/libkokkosalgorithms.a packages/kokkos/containers/src/libkokkoscontainers.a packages/tpetra/classic/LinAlg/libtpetraclassiclinalg.a packages/tpetra/classic/NodeAPI/libtpetraclassicnodeapi.a packages/tpetra/classic/src/libtpetraclassic.a packages/aztecoo/src/libaztecoo.a packages/triutils/src/libtriutils.a packages/epetra/src/libepetra.a packages/teuchos/kokkoscomm/src/libteuchoskokkoscomm.a packages/teuchos/kokkoscompat/src/libteuchoskokkoscompat.a packages/teuchos/remainder/src/libteuchosremainder.a packages/teuchos/numerics/src/libteuchosnumerics.a packages/teuchos/comm/src/libteuchoscomm.a packages/teuchos/parameterlist/src/libteuchosparameterlist.a packages/teuchos/parser/src/libteuchosparser.a /usr/lib/liblapack.dylib /usr/lib/libblas.dylib packages/teuchos/core/src/libteuchoscore.a packages/kokkos/core/src/libkokkoscore.a /usr/lib/libdl.dylib packages/zoltan/src/libzoltan.a -lm -lz -lcurl && :
```

**Comment by eric-c-cyr** [comment]
@krcb Its a globbing issue most likely. Run

```
touch packages/panzer/core/src/CMakeLists.txt
```

Hopefully that will trigger the cmake reconfigure.

**Comment by eric-c-cyr** [comment]
Alternatively I just modified the CMakeLists.txt file and it should auto reconfigure.

**Comment by jmgate** [comment]
> @eric-c-cyr @rppawlo It doesn't look like I'm authorized to checkout this PR (I'm not listed as a collaborator on this project).

@krcb, were you perhaps trying to check out `EM-Plasma/Trilinos:periodic-dg-problems` instead of `eric-c-cyr/Trilinos:periodic-dg-problems`?

**Comment by krcb** [comment]
@jmgate Since I don't have developer access EM-Plasma/Trilinos, I can't see the instructions on how to checkout the original branch. I just have a "You're not authorized" message. I'm also not authorized to access `eric-c-cyr/Trilinos:periodic-dg-problems`.

**Comment by jmgate** [comment]
Thanks for the feedback.  I'll try to figure out what's going on.

**Comment by eric-c-cyr** [comment]
Alright, I think this is ready for a review. @rppawlo would you mind looking at this?

**Comment by eric-c-cyr** [comment]
Alright, I think this is ready for a review. @rppawlo would you mind looking at this?

**Comment by jmgate** [comment]
@krcb, are you in CSRI sometime soon such that we could try to figure out what's going on?  As far as I can tell, EM-Plasma/Trilinos is a public fork of a public repo, so everyone should be default have read access.  I was able to confirm Scot Swan (can't find his GitHub username) can clone and we haven't done anything special to give him access, so I'm not sure what's going on.

File: packages/panzer/disc-fe/src/Panzer_IntegrationValues2.cpp [hunk 1]
```diff
@@ -495,6 +496,71 @@ convertNormalToRotationMatrix(const T normal[3], T transverse[3], T binormal[3])
 
 }
 
+template <typename Scalar>
+void IntegrationValues2<Scalar>::
+swapQuadraturePoints(int cell,
+                     int a,
+                     int b)
+{
```
[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply packages/panzer/disc-fe/src/Panzer_IntegrationValues2.cpp]: In fact this operates on all cells regardless of periodicity.

[the remark this answers is not in the document]
Comment (eric-c-cyr) [reply packages/panzer/disc-fe/src/Panzer_IntegrationValues2.cpp]: I agree that you are correct that the swap should happened on multiple cells at once...except this is done in the context of a sorting operation. I'm not sure it actually can without changing this to a parallel sort of some sort. If you look at reorder you can see why. Its a highly sequential sort.

**Comment by eric-c-cyr** [comment]
Closing to merge else where

**Pull request closed without merging**